File: backend/backend_main.py
from backend.langchain_chains.chains import run_simulation

class BackendMain:
    def __init__(self):
        self.history_dict = {}

    def simulate_year(self, year: str, current_event: str = None, user_decision: str = None) -> dict:
        if current_event is None:
            current_event = "The world stands on the brink of change; subtle signs of unrest are emerging globally."
        if user_decision is None:
            user_decision = "No decision made yet."
        
        simulation_result = run_simulation(year, current_event, user_decision)
        if not simulation_result.get("success"):
            print("Simulation error:", simulation_result.get("error"))
        else:
            print("\nInitial Simulation Output:")
            import json
            try:
                simulation_result = json.loads(simulation_result["data"])
                print(simulation_result)
            except json.JSONDecodeError as e:
                print("Error parsing JSON output:", e)
                print(simulation_result)

        self.history_dict[year] = simulation_result
        return simulation_result

    def get_history(self, year: str) -> dict:
        return self.history_dict.get(year, None)
    
    def get_all_history(self) -> dict:
        return self.history_dict

    
    
    

File: backend_ts/backendMain.ts
import { 
  combineSimulationInputs,
  buildHistoricalContextChain,
  runSimulation,
  parseSimulationOutput,
  generateAgentIdentity,
  chatWithUser,
  SimulationResult
} from "./chains";

export class Backend {
  private historyDict: Map<string, SimulationResult> = new Map();
  private currentYear: string | null = null;
  private latestSimulationResult: SimulationResult | null = null;

  async simulateYear(
    year: string,
    currentEvent?: string,
    userDecision?: string
  ): Promise<SimulationResult> {
    const result = await runSimulation(
      year,
      currentEvent || "The world stands on the brink of change...", 
      userDecision || "No decision made yet."
    );

    this.historyDict.set(year, result);
    this.currentYear = year;
    
    return result;
  }

  async getEventsByYear(year: string): Promise<Record<string, string[]>> {
    const result = this.historyDict.get(year);
    if (!result) throw new Error("No simulation result available");
    
    const eventsByRegion: Record<string, string[]> = {};
    for (const event of result.future_events) {
      const region = event.location || "Unknown";
      eventsByRegion[region] = eventsByRegion[region] || [];
      eventsByRegion[region].push(event.event_description);
    }
    return eventsByRegion;
  }

  async chatWithAgent(region: string, message: string): Promise<string> {
    if (!this.currentYear) throw new Error("No active simulation year");
    
    const identity = generateAgentIdentity(region);
    const quantities = latestSimulationResult?.regional_quantities?.[region];
    const regionalContext = quantities ? 
      `Quantities: ${JSON.stringify(quantities)}` : 
      "No regional data available";

    return chatWithUser(
      this.currentYear,
      region,
      message,
      regionalContext,
      identity
    );
  }
} 